pkg/diff/diff.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package diff
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/pkg/archive"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+const (
+	// ChangeTypeKey defines the key for the type of diff change stored in the tar Xattrs header
+	ChangeTypeKey = "change_type"
+	// ExcludeString defines the string that specifies a path exclusion.
+	ExcludeString = "exclude"
+	// IncludeString defines the string that specifies a path inclusion.
+	// Note: An included path that sits under an excluded path should be included.
+	// For example, if / is excluded, and /files is included, /files should  be added to the archive.
+	IncludeString = "include"
+	// RebaseString defines the path that was stripped when copying from a volume vmdk.
+	// This is necessary as the volume may be mounted deep in the container filesystem,
+	// but files on the volume are seen as living in the filesystem root when being diffed.
+	RebaseString = "rebase"
+)
+
+// TODO(jzt): move this to archive package?
+// PathSpec describes rules for handling specified paths during archival
+type PathSpec struct {
+	Inclusions map[string]struct{}
+	Exclusions map[string]struct{}
+	RebasePath string
+}
+
+// for sort.Sort
+type changesByPath []archive.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar archive containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := archive.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
+
+	spec, err := processPathSpec(op, pathspec)
+	if err != nil {
+		return nil, err
+	}
+
+	// prune excluded paths from change set
+	for i, change := range changes {
+		if skip(op, change.Path, spec) {
+			changes = append(changes[:i], changes[i+1:]...)
+		}
+	}
+
+	return Tar(op, newDir, changes, spec, data)
+}
+
+func Tar(op trace.Operation, dir string, changes []archive.Change, spec *PathSpec, data bool) (io.ReadCloser, error) {
+	var (
+		err error
+		hdr *tar.Header
+	)
+
+	r, w := io.Pipe()
```

Comment:
maybe a comment saying it is callers responsibility to close/handle errors of the read side of the pipe?